plugins/inputs/tacacs/tacacs_test.go

```diff
@@ -0,0 +1,255 @@
+	require.Equal(t, false, acc3.HasInt64Field("tacacs", "responsetime_ms"))
+
+	plugin = &Tacacs{
+		Servers:         []string{"localhost:9999"},
+		Username:        config.NewSecret([]byte(`testusername`)),
+		Password:        config.NewSecret([]byte(`testpassword`)),
+		Secret:          config.NewSecret([]byte(`testsecret`)),
+		RequestAddr:     "127.0.0.1",
+		ResponseTimeout: config.Duration(time.Nanosecond * 1),
+		Log:             testutil.Logger{},
+	}
+	var acc4 testutil.Accumulator
+	require.NoError(t, plugin.Init())
+	require.NoError(t, plugin.Gather(&acc4))
+	require.Len(t, acc4.Errors, 1)
+	require.ErrorContains(t, acc4.Errors[0], "error on new tacacs authentication start request to localhost:9999 : dial tcp")
+	require.Equal(t, false, acc4.HasTag("tacacs", "source"))
+	require.Equal(t, false, acc4.HasInt64Field("tacacs", "responsetime_ms"))
+}
+
+func TestTacacsIntegration(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration test in short mode")
+	}
+
+	container := testutil.Container{
+		Image:        "dchidell/docker-tacacs",
+		ExposedPorts: []string{"49/tcp"},
+		WaitingFor: wait.ForAll(
+			wait.ForLog("Starting server..."),
+		),
+	}
+	err := container.Start()
+	require.NoError(t, err, "failed to start container")
+	defer container.Terminate()
+
+	port := container.Ports["49"]
+
+	// Define the testset
+	var testset = []struct {
+		name           string
+		testingTimeout config.Duration
+		serverToTest   string
+		expectSuccess  bool
+		usedPassword   string
+	}{
+		{
+			name:           "timeout_3s",
+			testingTimeout: config.Duration(time.Second * 3),
+			serverToTest:   container.Address + ":" + port,
+			expectSuccess:  true,
+			usedPassword:   "cisco",
+		},
+		{
+			name:           "timeout_0s",
+			testingTimeout: config.Duration(0),
+			serverToTest:   container.Address + ":" + port,
+			expectSuccess:  true,
+			usedPassword:   "cisco",
+		},
+		{
+			name:           "wrong_pw",
+			testingTimeout: config.Duration(time.Second * 5),
+			serverToTest:   container.Address + ":" + port,
+			expectSuccess:  false,
+			usedPassword:   "wrongpass",
+		},
+	}
+
+	for _, tt := range testset {
+		t.Run(tt.name, func(t *testing.T) {
+			// Setup the plugin-under-test
+			plugin := &Tacacs{
+				ResponseTimeout: tt.testingTimeout,
+				Servers:         []string{tt.serverToTest},
+				Username:        config.NewSecret([]byte(`iosadmin`)),
+				Password:        config.NewSecret([]byte(tt.usedPassword)),
+				Secret:          config.NewSecret([]byte(`ciscotacacskey`)),
+				RequestAddr:     "127.0.0.1",
+				Log:             testutil.Logger{},
+			}
+			var acc testutil.Accumulator
+
+			// Startup the plugin
+			require.NoError(t, plugin.Init())
+
+			// Gather
+			require.NoError(t, plugin.Gather(&acc))
+
+			if tt.expectSuccess {
+				if len(acc.Errors) > 0 {
+					t.Errorf("error occured in test where should be none, error was: %s", acc.Errors[0].Error())
+				}
```

Comment:
```suggestion
				require.Lenf(t, acc.Errors, 0, "error occurred in test where should be none, error was: %s", acc.Errors[0].Error())
```

Reply:
changed it to use require.Lenf as suggested

Reply:
Reverted it, seems t.Errorf evaluates/constructs the string even if no error occurs (I would expect it to get constructed only when needed), causing index out of range

Reply:
Strange, then why does it work in the `if tt.name == "wrong_pw" {` section?

Reply:
There acc.Errors[0].Error() is not sent as a parameter (when no error exists, acc.Errors is empty, so no index 0 is present)

Reply:
Right, then it should be `acc.FirstError()` or simply this:
```suggestion
				require.NoError(t, acc.FirstError())
```

Reply:
Thats a good helper function (FirstError), changed it, will appear in next push (hopefully wont be many left)